src/Google.Cloud.Metadata.V1/MetadataClient.cs

```diff
@@ -0,0 +1,571 @@
+        /// </summary>
+        /// <remarks>
+        /// The project metadata contains information about the project, such as its ID, and custom metadata and it can be
+        /// accessed without authentication from an application running on an instance of that project. Check <see cref="IsServerAvailable"/> or
+        /// <see cref="IsServerAvailableAsync(CancellationToken)"/> to determine if the metadata can be accessed.
+        /// </remarks>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>A task containing the <see cref="InstanceMetadata"/> representing the metadata.</returns>
+        public virtual Task<ProjectMetadata> GetProjectMetadataAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        /// Gets the value indicating whether the metadata server or an emulator for the server is available synchronously.
+        /// </summary>
+        /// <remarks>
+        /// Normally, the metadata server can only be accessed from an application running on a Google Compute Engine VM instance or
+        /// Google App Engine Flexible Environment, but if an metadata server emulator is available, this will also return true.
+        /// </remarks>
+        /// <returns>True if the normal metadata server or an emulator is available; false otherwise.</returns>
+        public virtual bool IsServerAvailable()
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        /// Gets the value indicating whether the metadata server or an emulator for the server is available asynchronously.
+        /// </summary>
+        /// <remarks>
+        /// Normally, the metadata server can only be accessed from an application running on a Google Compute Engine VM instance or
+        /// Google App Engine Flexible Environment, but if an metadata server emulator is available, this will also return true.
+        /// </remarks>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>A task containing true if the normal metadata server or an emulator is available; false otherwise.</returns>
+        public virtual Task<bool> IsServerAvailableAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            throw new NotImplementedException();
+        }
+
+        // TODO: The http requests to modify the metadata require a fingerprint of the current metadata state to make sure you have the latest info
+        //       when making an update. This design does not allow for the same type of guarantee. Should we instead store the fingerprint on the
+        //       metadata classes internally and move these modification methods to InstanceMetadata and ProjectMetadata?
```

Comment:
I think it would make sense to store the fingerprint in the metadata classes, then fail with a known exception when the operation fails because the fingerprint has changed. We also discussed having a callback where the user can specify the changes they want and let the library retry as necessary, I think that having the fingerprint in the metadata would allow the user to do this logic if expected, specially if the value they want to set depends on other values present in the metadata.

If the value is a single value that is being changed, then we can hide this in the methods implemented in this class, by reading the metadata and effecting the changes as necessary ourselves.